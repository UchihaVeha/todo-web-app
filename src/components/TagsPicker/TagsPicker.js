// @flow
import * as React from 'react';
import { withStyles } from 'material-ui/styles';
import { Chip } from 'material-ui';
import { FormControl, FormHelperText, FormLabel } from 'material-ui/Form';

const styles = theme => ({
  chip: {
    marginRight: theme.spacing.unit,
    marginTop: theme.spacing.unit
  },
  wrapper: {
    width: '100%',
    display: 'flex',
    flexFlow: 'row wrap'
  },
  inputWrapper: {
    flexGrow: 1,
    marginTop: theme.spacing.unit * 2
  },
  input: {
    border: 0,
    width: '100%',
    height: 30,
    '&::-webkit-input-placeholder': {
      color: 'rgba(0, 0, 0, 0.3)'
    },
    '&:focus': {
      outline: 'none'
    }
  }
});

type Props = {
  tags: string[],
  error: ?string,
  onRemoveTag: (name: string) => void,
  onAddTag: (name: string) => void,
  classes: Return<typeof styles>
};

class TagsPicker extends React.Component<Props> {
  static defaultProps = {
    error: null
  };

  input: ?HTMLInputElement;

  handleBlur = ({ target }: SyntheticInputEvent<>) => {
    if (target.value !== '') {
      this.props.onAddTag(target.value);
      if (this.input) this.input.value = '';
    }
  };

  handleKeyPress = (event: SyntheticInputEvent<>): void => {
    if (
      (event.charCode === 13 || event.charCode === 32) &&
      event.target.value !== ''
    ) {
      this.props.onAddTag(event.target.value);
      if (this.input) this.input.value = '';
    }
  };

  render() {
    const { tags, classes, error } = this.props;
    return (
      <FormControl fullWidth margin="dense" error={!!error}>
        <FormLabel>Tags</FormLabel>
        <div className={classes.wrapper}>
          {tags.map(tag => (
            <Chip
              label={tag}
              className={classes.chip}
              key={tag}
              onRequestDelete={() => this.props.onRemoveTag(tag)}
            />
          ))}
          <div className={classes.inputWrapper}>
            <input
              ref={input => {
                this.input = input;
              }}
              className={classes.input}
              placeholder="Add Tag"
              onKeyPress={this.handleKeyPress}
              onBlur={this.handleBlur}
            />
          </div>
        </div>
        <FormHelperText>{error}</FormHelperText>
      </FormControl>
    );
  }
}

export default withStyles(styles, { name: 'TagsPicker' })(TagsPicker);
