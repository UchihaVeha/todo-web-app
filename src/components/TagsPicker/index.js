import TagsPicker from './TagsPicker';

export default TagsPicker;
