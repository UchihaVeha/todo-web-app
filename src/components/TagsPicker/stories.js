import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import ThemeProvider from 'themes/ThemeProvider';
import TagsPicker from './TagsPicker';

const tags = ['JQuery', 'Backbone', 'Marionette', 'Angular', 'React'];

storiesOf('TagsPicker', module)
  .addDecorator(story => (
    <div style={{ padding: 30, maxWidth: 600 }}>
      <ThemeProvider>{story()}</ThemeProvider>
    </div>
  ))
  .add('empty', () => (
    <TagsPicker
      tags={[]}
      onAddTag={action('ADD TAG')}
      onRemoveTag={action('REMOVE TAG')}
    />
  ))
  .add('with items', () => (
    <TagsPicker
      tags={tags}
      onAddTag={action('ADD TAG')}
      onRemoveTag={action('REMOVE TAG')}
    />
  ));
