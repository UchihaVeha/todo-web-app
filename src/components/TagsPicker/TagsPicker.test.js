import React from 'react';
import { shallow, mount } from 'enzyme';
import { FormHelperText, SvgIcon } from 'material-ui';
import TagsPicker from './TagsPicker';

describe('TagsPicker component', () => {
  let wrapper;

  beforeEach(() => {
    const tags = ['test tag'];
    wrapper = shallow(<TagsPicker tags={tags} />);
  });

  it('should render withStyles', () => {
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('should render component', () => {
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('should call onAddTag on key press', () => {
    const value = 'test value';
    const onAddTag = jest.fn();
    const input = shallow(<TagsPicker tags={[]} onAddTag={onAddTag} />)
      .dive()
      .find('input');
    input.simulate('keyPress', { charCode: 13, target: { value: '' } });
    expect(onAddTag).not.toHaveBeenCalled();
    input.simulate('keyPress', { charCode: 32, target: { value: '' } });
    expect(onAddTag).not.toHaveBeenCalled();
    input.simulate('keyPress', { charCode: 13, target: { value } });
    expect(onAddTag).toHaveBeenCalledWith(value);
    input.simulate('keyPress', { charCode: 32, target: { value } });
    expect(onAddTag).toHaveBeenCalledWith(value);
  });

  it('should call onAddTag on blur', () => {
    const onAddTag = jest.fn();
    const value = 'test value';
    wrapper = shallow(<TagsPicker tags={[]} onAddTag={onAddTag} />).dive();
    wrapper.find('input').simulate('blur', { target: { value: '' } });
    expect(onAddTag).not.toHaveBeenCalled();
    wrapper.find('input').simulate('blur', { target: { value } });
    expect(onAddTag).toHaveBeenCalledWith(value);
  });

  it('should call onRemoveTag', () => {
    const onRemoveTag = jest.fn();
    const tag = 'Test Tag';
    wrapper = mount(<TagsPicker tags={[tag]} onRemoveTag={onRemoveTag} />);
    wrapper.find(SvgIcon).simulate('click');
    expect(onRemoveTag).toHaveBeenCalledWith(tag);
  });

  it('should render error', () => {
    const error = 'test error';
    wrapper = shallow(<TagsPicker tags={[]} error={error} />).dive();
    expect(wrapper.dive().find(FormHelperText)).toHaveProp('children', error);
  });
});
