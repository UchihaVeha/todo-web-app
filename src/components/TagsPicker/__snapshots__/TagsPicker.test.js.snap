// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsPicker component should render component 1`] = `
<FormControl
  classes={
    Object {
      "fullWidth": "MuiFormControl-fullWidth-8",
      "marginDense": "MuiFormControl-marginDense-7",
      "marginNormal": "MuiFormControl-marginNormal-6",
      "root": "MuiFormControl-root-5",
    }
  }
  component="div"
  disabled={false}
  error={false}
  fullWidth={true}
  margin="dense"
  required={false}
>
  <withStyles(FormLabel)>
    Tags
  </withStyles(FormLabel)>
  <div
    className="TagsPicker-wrapper-2"
  >
    <withStyles(Chip)
      className="TagsPicker-chip-1"
      key="test tag"
      label="test tag"
      onRequestDelete={[Function]}
    />
    <div
      className="TagsPicker-inputWrapper-3"
    >
      <input
        className="TagsPicker-input-4"
        onBlur={[Function]}
        onKeyPress={[Function]}
        placeholder="Add Tag"
      />
    </div>
  </div>
  <withStyles(FormHelperText) />
</FormControl>
`;

exports[`TagsPicker component should render withStyles 1`] = `
<withStyles(FormControl)
  error={false}
  fullWidth={true}
  margin="dense"
>
  <withStyles(FormLabel)>
    Tags
  </withStyles(FormLabel)>
  <div
    className="TagsPicker-wrapper-2"
  >
    <withStyles(Chip)
      className="TagsPicker-chip-1"
      key="test tag"
      label="test tag"
      onRequestDelete={[Function]}
    />
    <div
      className="TagsPicker-inputWrapper-3"
    >
      <input
        className="TagsPicker-input-4"
        onBlur={[Function]}
        onKeyPress={[Function]}
        placeholder="Add Tag"
      />
    </div>
  </div>
  <withStyles(FormHelperText) />
</withStyles(FormControl)>
`;
